barter-data/src/exchange/coinbase/book/l2.rs

```diff
@@ -0,0 +1,665 @@
+use std::fmt::Debug;
+use std::future::Future;
+use async_trait::async_trait;
+use chrono::{DateTime, Utc};
+use derive_more::Constructor;
+use futures_util::future::try_join_all;
+use reqwest::Client;
+use rust_decimal::Decimal;
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc::UnboundedSender;
+use barter_instrument::exchange::ExchangeId;
+use barter_instrument::Side;
+use barter_integration::error::SocketError;
+use barter_integration::protocol::websocket::WsMessage;
+use barter_integration::subscription::SubscriptionId;
+use barter_integration::Transformer;
+use crate::books::{Level, OrderBook};
+use crate::error::DataError;
+use crate::event::{MarketEvent, MarketIter};
+use crate::exchange::coinbase::Coinbase;
+use crate::exchange::Connector;
+use crate::{Identifier, SnapshotFetcher};
+use crate::exchange::coinbase::market::CoinbaseMarket;
+use crate::instrument::InstrumentData;
+use crate::subscription::book::{OrderBookEvent, OrderBooksL2};
+use crate::subscription::{Map, Subscription};
+use crate::transformer::ExchangeTransformer;
+
+pub const HTTP_COINBASE_API: &str = "https://api.exchange.coinbase.com";
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CoinbaseOrderBookSnapshot {
+    pub sequence: u64,
+    pub bids: Vec<CoinbaseOrderBookSnapshotLevel>,
+    pub asks: Vec<CoinbaseOrderBookSnapshotLevel>,
+    #[serde(
+        alias = "time"
+    )]
+    pub time_exchange: DateTime<Utc>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CoinbaseOrderBookL2Snapshot {
+    pub product_id: String,
+    pub bids: Vec<CoinbaseOrderBookL2SnapshotLevel>,
+    pub asks: Vec<CoinbaseOrderBookL2SnapshotLevel>,
+}
```

Comment:
product_id can probably be a &str: 
```suggestion
pub struct CoinbaseOrderBookL2Snapshot<'a {
    pub product_id: &'a str,
    pub bids: Vec<CoinbaseOrderBookL2SnapshotLevel>,
    pub asks: Vec<CoinbaseOrderBookL2SnapshotLevel>,
}
```